internal/controller/buildpipeline/buildpipeline_adapter.go

```diff
@@ -305,3 +266,68 @@ func (a *Adapter) updateBuildPipelineRunWithFinalInfo(canRemoveFinalizer bool) e
 		return nil
 	})
 }
+
+// checkForSnapshotsCount makes sure that only one snapshot is associated with build pipelinerun and annotate that PLR with snapshot
+// informs about fact when PLR is associated with more existing snapshots
+func (a *Adapter) checkForSnapshotsCount(canRemoveFinalizer *bool, existingSnapshots *[]applicationapiv1alpha1.Snapshot) (result controller.OperationResult, err error) {
+	if len(*existingSnapshots) == 1 {
+		existingSnapshot := (*existingSnapshots)[0]
+		a.logger.Info("There is an existing Snapshot associated with this build pipelineRun, but the pipelineRun is not yet annotated",
+			"snapshot.Name", existingSnapshot.Name)
+		err := a.annotateBuildPipelineRunWithSnapshot(&existingSnapshot)
+		if err != nil {
+			a.logger.Error(err, "Failed to update the build pipelineRun with snapshot name",
+				"pipelineRun.Name", a.pipelineRun.Name)
+			return controller.RequeueWithError(err)
+		}
+	} else {
+		a.logger.Info("The build pipelineRun is already associated with more than one existing Snapshot")
+	}
+	*canRemoveFinalizer = true
+	return controller.ContinueProcessing()
+}
+
+// failedOrDeletedPLR checks for pipelinerun state and proceeds according to it,
+// failed or in running state > report this into a logger and set canRemoveFinalizer flag to true
+func (a *Adapter) failedOrDeletedPLR(canRemoveFinalizer *bool) (result controller.OperationResult, err error) {
+	// The build pipelinerun has not finished
+	if h.HasPipelineRunFinished(a.pipelineRun) || a.pipelineRun.GetDeletionTimestamp() != nil {
+		// The pipeline run has failed
+		// OR pipeline has been deleted but it's still in running state (tekton bug/feature?)
+		a.logger.Info("Finished processing of unsuccessful build PLR",
+			"statusCondition", a.pipelineRun.GetStatusCondition(),
+			"deletionTimestamp", a.pipelineRun.GetDeletionTimestamp(),
+		)
+		*canRemoveFinalizer = true
+		return controller.ContinueProcessing()
+	}
+	return controller.ContinueProcessing()
+}
+
+// failedToCreateSnapshot stops reconcilation immediately when snapshot cannot be created
+func (a *Adapter) failedToCreateSnapshot(canRemoveFinalizer *bool, cerr error) (result controller.OperationResult, err error) {
+	a.logger.Error(cerr, "Failed to create Snapshot")
+	if errors.IsForbidden(cerr) {
+		// we cannot create a snapshot (possibly because the snapshot quota is hit) and we don't want to block resources, user has to retry
+		*canRemoveFinalizer = true
+		return controller.StopProcessing()
+	}
+	return controller.RequeueWithError(cerr)
+}
+
+// plrWithCustomizedError checks for customized error returned by PipelineRun result,
+// updates build PipelineRun annotation with this error and exits
+func (a *Adapter) plrWithCustomizedError(canRemoveFinalizer *bool, cerr error, context context.Context, pipelineRun *tektonv1.PipelineRun, client client.Client, logger h.IntegrationLogger) (result controller.OperationResult, err error) {
```

Comment:
I think renaming this function to `pipelineRunWithCustomizedError` would make it easier to read when going through the code.

Reply:
Agreed, renamed.